tools/java/org/unicode/cldr/util/XMLSource.java

```diff
@@ -1747,4 +1764,711 @@ public DtdType getDtdType() {
+                    }
+                }
+            }
+            value = trimWhitespaceSpecial(value);
+            source.add(fullXPath, value);
+        }
+
+        private void pop(String qName) {
+            Log.logln(LOG_PROGRESS, "pop\t" + qName);
+            --level;
+
+            if (lastChars.length() != 0 || justPopped == false) {
+                boolean acceptItem = minimalDraftStatus == DraftStatus.unconfirmed;
+                if (!acceptItem) {
+                    if (draftMatcher.reset(currentFullXPath).find()) {
+                        DraftStatus foundStatus = DraftStatus.valueOf(draftMatcher.group(1));
+                        if (minimalDraftStatus.compareTo(foundStatus) <= 0) {
+                            // what we found is greater than or equal to our status
+                            acceptItem = true;
+                        }
+                    } else {
+                        acceptItem = true; // if not found, then the draft status is approved, so it is always ok
+                    }
+                }
+                if (acceptItem) {
+                    // Change any deprecated orientation attributes into values
+                    // for backwards compatibility.
+                    boolean skipAdd = false;
+                    if (currentFullXPath.startsWith("//ldml/layout/orientation")) {
+                        XPathParts parts = XPathParts.getFrozenInstance(currentFullXPath);
+                        String value = parts.getAttributeValue(-1, "characters");
+                        if (value != null) {
+                            addPath("//ldml/layout/orientation/characterOrder", value);
+                            skipAdd = true;
+                        }
+                        value = parts.getAttributeValue(-1, "lines");
+                        if (value != null) {
+                            addPath("//ldml/layout/orientation/lineOrder", value);
+                            skipAdd = true;
+                        }
+                    }
+                    if (!skipAdd) {
+                        addPath(currentFullXPath, lastChars);
+                    }
+                    lastLeafNode = lastActiveLeafNode = currentFullXPath;
+                }
+                lastChars = "";
+            } else {
+                Log.logln(LOG_PROGRESS && lastActiveLeafNode != null, "pop: zeroing last leafNode: "
+                    + lastActiveLeafNode);
+                lastActiveLeafNode = null;
+                if (comment != null) {
+                    source.addComment(lastLeafNode, comment, XPathParts.Comments.CommentType.POSTBLOCK);
+                    comment = null;
+                }
+            }
+            // currentXPath = stripAfter(currentXPath, qName);
+            currentFullXPath = stripAfter(currentFullXPath, qName);
+            justPopped = true;
+        }
+
+        static Pattern WHITESPACE_WITH_LF = PatternCache.get("\\s*\\u000a\\s*");
```

Comment:
`static final` ?
And move all 3 of these at the top of the class.

Reply:
Pattern can be moved, and the UnicodeSet moved (AND)

new UnicodeSet("[:cc:]").freeze();

Matcher shouldn't be static. Is mutable.

Reply:
Will change